README.md
# dart-mc-launcher v0.0.1

dart-mc-launcher is a framework for starting minecraft clients and servers from dart

## Installation

```yaml
name: [unique_namespace]
dependencies:  
  mc_launcher: ^0.0.1
```

Also remember to replace the `[unique_namespace]` with your own project name and run

```shell
$ pub get
```

if you are using a clever editor like `vscode`, you can propably you can propably skip this step

## Getting started

### **Base**

First import the package

```dart
import 'package:mc_launcher/core.dart';
```

The hole package is written asynchronously, so we will use an async main method to demonstarte the package functionality.

```dart
void main(List<String> args) async {
}
```

### **Server**

Inside of it we will create a simple vanilla server the subdirectory server. This directory will be created automatically. Also we will set the default eula value to true, so we can start the server. It is defaultly false, so we have to accept it manually, but with this, our server is directly ready to start.
I set the version to 1.14.4 here, if you do not set the version, it will automatically use the latest version

```dart
MinecraftServer server = await VanillaServer.create("./server", defaultEula: true, version: "1.14.4");
```

Now we start the server

```dart
server.start();
```

We can execute Commands using

```dart
server.executeCommand("say hello world :)");
```

And stop it using

```dart
server.stop();
```

more configurations you can set like this:

```dart
server.properties.mar
```

### **Client**

To create clients also is really easy. You just have to specify the version, the root directory. If you want you can also specify maximum *(default: 1024)* and minimum *(default 512)* ram. *It will take some time for the client to download the assets at this point on the first start*

```dart
MinecraftClient client = await VanillaClient.createClient(rootDir: "./client", version: "1.14.4", maxRam: 6000, minRam: 4000);
```

To start the clients just use client.start.

```dart
client.start()
```

you can **not** use .stop() on the clients, you can just kill them via client.kill. The problem with killing the client is, that it saves no worlds, etc when you kill them, so be carefull with this function

```dart
client.kill()
```

## License

This package is licensed under the [BSD2-Clause](https://github.com/MinimineLP/dart-mc-launcher/blob/master/LICENSE)

## Authors

- **Minimine** - *Initial work* - [Minimine](https://github.com/MinimineLP)

## Issues

Please report bugs, issues, feature request and stuff like that on the [github issue tracker](https://github.com/MinimineLP/dart-mc-launcher/issues)

If you are coding yourself and have a fix for my code feel free do send a [pull request](https://github.com/MinimineLP/dart-mc-launcher/pulls)


CHANGELOG.md
# Changelog

## 0.0.1

- Implemented Vanilla client + server handler
- Implemented Fabric client + server handler
- Implemented properties file parser & encoder
- Implemented custom console progress bar


lib/vanilla/server.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:mc_launcher/java.dart';
import 'package:mc_launcher/minecraft/file_creator.dart';
import 'package:mc_launcher/minecraft/server.dart';
import 'package:mc_launcher/minecraft/server_properties.dart';
import 'package:meta/meta.dart';
import 'package:http/http.dart' as http;
import 'package:mc_launcher/properties_format.dart';
import 'package:mc_launcher/utils.dart';
import 'package:path/path.dart' as path;

class VanillaServer implements MinecraftServer {
  
  final String serverdir;
  final DartMcLauncherProperties properties;
  final ServerProperties server_properties;
  Process process;
  bool get running => _running;

  List<void Function(String)> onStdOut = [
    (data) { 
      stdout.write(data);
    }
  ];

  List<void Function(String)> onStdErr = [
    (data) { 
      stderr.write(data);
    }
  ];

  bool _running = false;

  VanillaServer._init(this.serverdir, this.properties, this.server_properties);
  
  static Future<VanillaServer> install(String serverdir, {String version, bool eula = false}) async {
  
    await createDirIfNotExists(serverdir);
  
    http.Response response = await http.get('https://launchermeta.mojang.com/mc/game/version_manifest.json');
    Map<String,dynamic> version_manifest = json.decode(response.body);
    if(version == null) version = version_manifest["latest"]["release"];
      
    _VersionInformations informations;
  
    for(Map<String,dynamic> value in version_manifest["versions"]) {
      if(value["id"] != version) continue;
      http.Response response = await http.get(value["url"]);
      Map<String,dynamic> versioninformation = json.decode(response.body);
      informations = _VersionInformations(id: value["id"],releaseTime: value["releaseTime"], time: value["time"], type: value["time"], url: value["url"], download: versioninformation["downloads"]["server"]["url"]);
      break;
    }
  
    DartMcLauncherVanillaProperties properties = new DartMcLauncherVanillaProperties(jar_file: "server.jar", maxram: 1024, minram: 512, minecraft_version: version, gui: false, file: "$serverdir/dart-mc-launcher.properties");
    ServerProperties serverProperties = new ServerProperties(file: '$serverdir/server.properties');
    List<Future> futures = [
      properties.save(),
      downloadFile(informations.download, '$serverdir/server.jar'),
      createEula(serverdir, eula),
      serverProperties.save()
    ];
    futures.addAll(create_start_files(serverdir, 'server.jar'));
  
    await Future.wait(futures);
    return VanillaServer._init(serverdir, properties, serverProperties);
  }
  
  static Future<VanillaServer> create(String serverdir, {String version, bool defaultEula = false}) async {
  
    DartMcLauncherVanillaProperties properties;
  
    if(!(await new Directory(serverdir).existsSync())
      || !(await new File(path.join(serverdir,"server.properties")).exists())) {
      return await install(serverdir, version: version, eula: defaultEula);
    } 
    else if(await new File(path.join(serverdir,"dart-mc-launcher.properties")).exists()) {
      properties = await DartMcLauncherVanillaProperties.loadFile(path.join(serverdir,"dart-mc-launcher.properties"));
    }
    else {
      throw("The server you are trying to load has no \"dart-mc-launcher.properties\" file, for that reason it can't be read by dart-mc-launcher. If you wanna create a new Server at this directory delete all contents, in other case you have to create a file manually");
    }
  
    return new VanillaServer._init(serverdir, properties, await ServerProperties.loadFile(path.join(serverdir,"server.properties")));
  }

  @override
  Future<VanillaServer> start() async {
    if(this.running) throw("Server already running!");
    this._running = true;

    List<String> args = ["-Xms${properties.minram}", "-Xmx${properties.maxram}", "-jar",properties.jar_file];
    if(!properties.gui) args.add("nogui");

    //print('$JAVA ' + args.join(" "));

    Completer<VanillaServer> completer = new Completer<VanillaServer>();

    this.process = await Process.start('$JAVA', args, workingDirectory: this.serverdir);

    process.stdout.transform(utf8.decoder).listen((data) { 
      this.onStdOut.forEach((e) {
        e(data);
      });
      if((!completer.isCompleted) && new RegExp(r'''Done \([1-9]+\.[1-9]+\w\)! For help, type "help"''').hasMatch(data)) {
        completer.complete(this);
      }
    });
    
    process.exitCode.then((v) {
      this._running = false;
      this.process = null;
    });

    process.stderr.transform(utf8.decoder).listen((data) { stderr.write(data); });

    return await completer.future;
  }

  @override
  Future<VanillaServer> stop() async {
    if(!this.running) throw("Server is not running!");
    executeCommand("stop");
    await process.exitCode;
    this._running = false;
    this.process = null;
    return this;
  }
  
  @override
  Future<VanillaServer> executeCommand(String cmd) async {
    if(!this.running) throw("Server is not running!");
    print("> $cmd");
    process.stdin.writeln(cmd);
    return this;
  }
}

class _VersionInformations {
  final String id;
  final String type;
  final String url;
  final String time;
  final String releaseTime;
  final String download;
  _VersionInformations({this.id,this.time, this.releaseTime,this.type,this.url,this.download});
}

class DartMcLauncherVanillaProperties implements DartMcLauncherProperties {
  PropertiesTemplate template = null;
  String minecraft_version;
  String jar_file;
  int minram;
  int maxram;
  bool gui = false;
  String file;

  DartMcLauncherVanillaProperties({
    @required this.minecraft_version,
    @required this.jar_file,
    @required this.maxram,
    @required this.minram,
    this.gui = false,

    /// The properies's file path
    @required this.file
  });

  DartMcLauncherVanillaProperties.loadMap(Map<String,dynamic> map) {
    this._loadMap(map);
  }

  DartMcLauncherVanillaProperties.loadProperties(Properties properties, this.file) {
    this._loadMap(properties.toMap());
    this.template = properties.toTemplate();
  }

  DartMcLauncherVanillaProperties.loadFile(String file): this.loadProperties(Properties.decode(File(file).readAsStringSync()), file);

  void _loadMap(Map<String,dynamic> map) {
      if(map["type"] != "vanilla") throw("Loading non-vanilla configuration with vanilla, type is \"${map["type"]}\"");
      this.minecraft_version = map["minecraft-version"];
      this.jar_file = map["jar-file"];
      this.minram = map["min-ram"];
      this.maxram = map["max-ram"];
      this.gui = map["gui"];
  }

  @override
  Map<String,dynamic> toMap() {
    return {
      "type": "vanilla",
      "minecraft-version": minecraft_version,
      "jar-file": jar_file,
      "min-ram": minram,
      "max-ram": maxram,
      "gui": gui,
    };
  }

  @override
  Properties toProperties() {
    return (template ?? new PropertiesTemplate([])).create(toMap());
  }

  @override
  Future<File> saveToFile(String file) {
    return toProperties().saveToFile(file);
  }

  @override
  void saveToFileSync(String file) {
    toProperties().saveToFileSync(file);
  }

  @override
  Future<File> save() {
    return saveToFile(file);
  }

  @override
  void saveSync() {
    saveToFileSync(file);
  }

  @override
  String toString() {
    return toProperties().toString();
  }
}